Migrate `ec_deployment`

dimuon opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

## Description
<!--- Describe your changes in detail. -->

**`ec_deployment` resource migration**

*Limitations.*
State upgrade is not currently implemented.

Few acceptance tests fail. Please see the section below.

## Related Issues
<!--- This project only accepts pull requests related to open issues. -->
<!--- If suggesting a new feature or change, please discuss it in an -->
<!--- issue first.  If fixing a bug, there should be an issue describing -->
<!--- it with steps to reproduce.  Please link to the any related issues -->
<!--- here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

* unit tests (majority of the relevant unit tests were migrated)
* acceptance tests
* manual testing

There are still few acceptance tests that fail:
- there are 4 acceptance tests (their names mention version 0.4.1) that verify migration of other data sources and resources from 0.4.1 to 0.6.0. To fix them we need implement a state upgrade for `ec_deployment`.
- `TestAccDeploymentExtension_pluginDownload` fails on 0.4.1 with the same error ("Custom plugin is not allowed at your subscription level.")
- `TestAccDeploymentExtension_bundleFile` fails quite often on both 0.4.1 and the branch with the "failed to upload file" error.
- `TestAccDeployment_withExtension` fails quite often with the same error as `TestAccDeploymentExtension_bundleFile`.

<!--- Please describe in detail how you tested your changes.  Include -->
<!--- details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in -->
<!--- all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improves code quality but has no user-facing effect)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation

## Readiness Checklist
<!--- Go over all the following points, and put an `x` in all the boxes -->
<!--- that apply.  If you're unsure about any of these, don't hesitate -->
<!--- to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

dimuon committed:
ec_deployment new schema

dimuon committed:
`ec_deployment` resource: explicit `id` attribute

dimuon committed:
`ec_deployment` schema: kibana block

dimuon committed:
`ec_deployment` schema: enterprise search, integration server, apm

dimuon committed:
ec_deployment new schemamodifications

dimuon commented:
@pascal-hofmann , the schema for `ec_deploment` resource is more or less complete. More probably we will need to tune/fix it once CRUD operations are ready. As for now I tested it just on a small set of examples.

dimuon committed:
TODO comments

dimuon committed:
Deployment resource data

pascal-hofmann commented on ec/internal/validators/urlvalidator.go:
```diff
@@ -79,5 +97,5 @@ func (v isURLWithSchemeValidator) Validate(ctx context.Context, req tfsdk.Valida
 
 func IsURLWithSchemeValidator(validSchemes []string) tfsdk.AttributeValidator {
-	return isURLWithSchemeValidator{}
+	return isURLWithSchemeValidator{ValidSchemes: []string{"http", "https"}}
```
```suggestion
	return isURLWithSchemeValidator{ValidSchemes: validSchemes}
```

pascal-hofmann commented on ec/internal/validators/oneOf.go:
```diff
@@ -0,0 +1,71 @@
+	}
+}
+
+// Known returns an AttributeValidator which ensures that any configured
+// attribute value:
+//
+//   - Is known.
```
```suggestion
// OneOf returns an AttributeValidator which ensures that any configured
// attribute value:
//
//   - Is one of the accepted values.
```

pascal-hofmann commented on ec/internal/validators/oneOf.go:
```diff
@@ -0,0 +1,71 @@
+//
+//   - Is known.
+//
+// Null (unconfigured) values are skipped.
```
```suggestion
```

pascal-hofmann commented on ec/internal/validators/notempty.go:
```diff
@@ -0,0 +1,58 @@
+	}
+}
+
+// Known returns an AttributeValidator which ensures that any configured
+// attribute value:
+//
+//   - Is known.
+//
+// Null (unconfigured) values are skipped.
+func NotEmpty() tfsdk.AttributeValidator {
```
```suggestion
// NotEmpty returns an AttributeValidator which ensures that any configured
// attribute value:
//
//   - Is known.
//   - Is set.
//   - Is no empty string.
//
func NotEmpty() tfsdk.AttributeValidator {
```

pascal-hofmann commented on ec/internal/validators/length.go:
```diff
@@ -0,0 +1,71 @@
+	}
+}
+
+// Known returns an AttributeValidator which ensures that any configured
+// attribute value:
+//
+//   - Is known.
+//
+// Null (unconfigured) values are skipped.
```
```suggestion
// Length returns an AttributeValidator which ensures that any configured
// attribute value:
//
//   - Has a length between min and max.
//
```

dimuon committed:
Update ec/internal/validators/oneOf.go

Co-authored-by: Pascal Hofmann <mail@pascalhofmann.de>

dimuon committed:
Update ec/internal/validators/oneOf.go

Co-authored-by: Pascal Hofmann <mail@pascalhofmann.de>

dimuon committed:
Update ec/internal/validators/notempty.go

Co-authored-by: Pascal Hofmann <mail@pascalhofmann.de>

dimuon committed:
Update ec/internal/validators/length.go

Co-authored-by: Pascal Hofmann <mail@pascalhofmann.de>

pascal-hofmann commented:
@dimuon Do you need help with ec_deployment or should I start with porting one of the other resources?

dimuon commented:
@pascal-hofmann , I think it makes sense to either start porting other resources or maybe adding acceptance tests for the migration, as recommended by Terraform - https://github.com/hashicorp/terraform-plugin-framework/blob/main/website/docs/plugin/framework/migrating/testing.mdx.

pascal-hofmann commented on ec/internal/validators/length.go:
```diff
@@ -0,0 +1,70 @@
+		return
+	}
+
+	if length := len(req.AttributeConfig.String()); length < v.min || length > v.max {
```
```suggestion
	if length := len(req.AttributeConfig.Value)); length < v.min || length > v.max {
```

Because:
```
// String returns a human-readable representation of the String value.
// The string returned here is not protected by any compatibility guarantees,
// and is intended for logging and error reporting.
func (s String) String() string {
	if s.Unknown {
		return attr.UnknownValueString
	}

	if s.Null {
		return attr.NullValueString
	}

	return fmt.Sprintf("%q", s.Value)
}
```



>%q	a double-quoted string safely escaped with Go syntax
- according to https://pkg.go.dev/fmt

pascal-hofmann commented on ec/internal/validators/notempty.go:
```diff
@@ -0,0 +1,59 @@
+
+// Validate runs the main validation logic of the validator, reading configuration data out of `req` and updating `resp` with diagnostics.
+func (v notEmptyValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
+	if req.AttributeConfig.IsUnknown() || req.AttributeConfig.IsNull() || req.AttributeConfig.String() == "" {
```
```suggestion
	if req.AttributeConfig.IsUnknown() || req.AttributeConfig.IsNull() || req.AttributeConfig.Value == "" {
```

pascal-hofmann commented on ec/internal/validators/oneOf.go:
```diff
@@ -0,0 +1,70 @@
+		return
+	}
+
+	if value := req.AttributeConfig.String(); !slice.HasString(v.values, value) {
```
```suggestion
	if value := req.AttributeConfig.Value; !slice.HasString(v.values, value) {
```

dimuon committed:
Update ec/internal/validators/length.go

Co-authored-by: Pascal Hofmann <mail@pascalhofmann.de>

dimuon committed:
Update ec/internal/validators/notempty.go

Co-authored-by: Pascal Hofmann <mail@pascalhofmann.de>

dimuon committed:
Update ec/internal/validators/oneOf.go

Co-authored-by: Pascal Hofmann <mail@pascalhofmann.de>

dimuon committed:
: read (WIP)

dimuon committed:
: read (WIP)

dimuon commented:
@pascal-hofmann , I've pushed still unfinished implementation of `Read` for `ec_deployment`. I tested it on a resource that was created with the previous version. The problem is that the plan is not empty for a configuration that doesn't introduce any change (basically I've created a resource using the old provider and then executed `terraform plan` using the new implementation and the existing resource configuration that was untouched). I'm looking into this. 

Also, it looks like changes you've suggested aren't compilable - `Value` property isn't available in the snippets you've commented on.

pascal-hofmann commented:
> Also, it looks like changes you've suggested aren't compilable - `Value` property isn't available in the snippets you've commented on.

My IDE (goland) deceived me with this. I had problems with Value.String() and assumed this is the same here, because goland found these usages. I should have tested my suggestions locally. Sorry for the confusion.

dimuon committed:
Fix tags implementation

dimuon committed:
playing with deployment resource definition

dimuon committed:
playing with deployment resource definition

dimuon committed:
Set MinItems for Elasticsearch block

dimuon committed:
Adding more logic to ec_deployment

dimuon committed:
Adding autoscaling and config to Elasticsearch topology

dimuon committed:
Add more fields to Elasticsearch

dimuon committed:
adding more fields to deployment resource

dimuon committed:
Refactoring

dimuon committed:
add kibana to deployment resource

dimuon committed:
add apm to deployment resource

dimuon committed:
integrations server in deployment resource

dimuon committed:
enterprise search in deployment resource

dimuon committed:
add more resources to deployment

dimuon committed:
traffic filters and observability in deployment resource

dimuon committed:
minor change

dimuon committed:
renaming

dimuon committed:
Fix for non-empty plan for some fields

dimuon committed:
minor fix

dimuon committed:
Move to the framework v0.12.0

dimuon committed:
renaming

pascal-hofmann commented on ec/ecresource-tpf/deploymentresource/deployment.go:
```diff
@@ -0,0 +1,146 @@
+	ApmSecretToken        types.String          `tfsdk:"apm_secret_token"`
+	TrafficFilter         []string              `tfsdk:"traffic_filter"`
+	Tags                  types.Map             `tfsdk:"tags"`
+	Elasticsearch         []*Elasticsearch      `tfsdk:"elasticsearch"`
+	Kibana                []*Kibana             `tfsdk:"kibana"`
+	Apm                   []*Apm                `tfsdk:"apm"`
+	IntegrationsServer    []*IntegrationsServer `tfsdk:"integrations_server"`
+	EnterpriseSearch      []*EnterpriseSearch   `tfsdk:"enterprise_search"`
+	Observability         []*Observability      `tfsdk:"observability"`
```
>The recommended way to deal with configuration or plan data for now is using types package types in Go structs, e.g.
> ```
> type myModel struct {
>   ExampleList types.List `tfsdk:"example_list"`
> }
> ```
-- https://discuss.hashicorp.com/t/terraform-plugin-framework-value-conversion-error-unhandled-unknown-value/44382/3

dimuon replied:
Yeah, it's a valid concern. It looks like there is also an option to leverage [appropriate interfaces](https://www.terraform.io/plugin/framework/accessing-values#detected-interfaces). 
I'll reconsider these definitions when I complete `Create`/`Update`.
Thank you for the heads-up.

dimuon committed:
Add payload methods to some deployment models

dimuon committed:
Add payload methods to some deployment models

dimuon committed:
add dep

dimuon committed:
deployment resource Create

dimuon committed:
fix compilation

dimuon committed:
Rewrite read and payload methods

dimuon committed:
delete state in create if deployment isn't found

dimuon committed:
refactor reading data from backend

dimuon committed:
refactor ec_deployment

dimuon committed:
small fixes

dimuon committed:
delete for ec_deployment

dimuon committed:
fix traffic filter rules deletion

dimuon committed:
update ec_deployment

dimuon committed:
minor fix

dimuon committed:
test for elasticsearch read

dimuon committed:
tests for elasticsearch

dimuon committed:
test for remote clusters

dimuon committed:
test for enterprise search

dimuon committed:
test for enterprise search read

dimuon committed:
increase unit test timeout and use default for 'parallel' setting

dimuon committed:
New unit test for ec_deployment (WIP)

dimuon committed:
new files

dimuon committed:
refactoring + tests

dimuon committed:
fix lint errors

dimuon committed:
delete obsolete file

dimuon committed:
unit tests for ec_deployment payload + fixes

dimuon committed:
another unit test

dimuon committed:
unit test for update ec_deployment (WIP)

dimuon committed:
unit test for ec_deployment update (WIP)

dimuon committed:
unit tests for update ec_deployment + fixes

dimuon committed:
unit tests for update ec_deployment + fixes

dimuon committed:
fix lint errors

dimuon committed:
update Framework to the latest version (0.14)

dimuon committed:
import for ec_deployment

dimuon committed:
minor fixes

dimuon committed:
make ec_deployment schema backward compatible

dimuon committed:
fix examples

dimuon committed:
fix docs

dimuon committed:
fix examples

dimuon committed:
fix examples

dimuon committed:
fix lint error

dimuon committed:
fix lint error

dimuon committed:
update imports

dimuon committed:
update lint and terrafmt versions

dimuon committed:
fixes

dimuon committed:
unit test for integrations_server payload

dimuon committed:
unit test for read integrations_server

dimuon committed:
replace blocks with attributes in acceptance tests

dimuon committed:
unit test for kibana payload

dimuon committed:
unit test for kibana read

dimuon committed:
unit tests for observability

dimuon committed:
unit test for deploymentNotFound

dimuon committed:
unit tests for stopped resources

dimuon committed:
unit tests for traffic filters

dimuon committed:
unit tests for getters and parsing credentials

dimuon committed:
unit test for testutils

dimuon committed:
unit tests for read deployments

dimuon committed:
unit tests for read deployments (WIP)

dimuon committed:
unit tests for read deployments (WIP)

dimuon committed:
unit tests for read deployments (WIP)

dimuon committed:
unit tests for read deployments (WIP)

dimuon committed:
unit tests for read deployments (WIP)

dimuon committed:
unit tests for read deployments (WIP)

dimuon committed:
unit tests for read deployments (WIP)

dimuon committed:
unit tests for read deployments

dimuon committed:
fixes

dimuon committed:
plan modifier UseStateIncludingNullForUnknown

dimuon committed:
remove obsolete files

dimuon committed:
fix Makefile
fix acceptance tests (WIP)
fix security acceptance test

dimuon committed:
fix observability acceptance test

dimuon committed:
minor fixes

dimuon committed:
fixes for acceptance tests

dimuon committed:
BREAKING: use blocks for elasticsearch and kibana

dimuon committed:
Revert "BREAKING: use blocks for elasticsearch and kibana"

This reverts commit ffa80a6bcdd36b059fb240049219fd1e79704169.

dimuon committed:
disable some plan modifiers

dimuon committed:
remove legacy provider

dimuon committed:
renamimng

dimuon committed:
blocks for elasticsearch and topology. failing test

dimuon committed:
Revert "blocks for elasticsearch and topology. failing test"

This reverts commit a45ba3be3b5a3480681b3e30354b3685e505fc23.

dimuon committed:
fix name

dimuon committed:
reproduce the missing data problem in a unit test

dimuon committed:
fix missing default value in the failing unit test

dimuon committed:
update test data

kuisathaverat committed:
fix: remove strategy DiffSuppressFunc (#539)

Commit:
Update CHANGELOG.md and NOTICE

dimuon committed:
add versioning to ec_deployment

dimuon committed:
fix path to test data in unit test

dimuon committed:
fix unit test

dimuon committed:
fix path to test data in unit test

dimuon committed:
delete obsolete code

dimuon committed:
refactoring

dimuon committed:
ec_deployment v2

dimuon committed:
add v2 for all models

dimuon committed:
v2 for observability

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
Fix the failing test by removing Computed from the ES topology.

+ some minor fixes

dimuon committed:
remove id attribute from ES topology

dimuon committed:
remove computed flag from other apps topology schemas

dimuon committed:
fix deployment_basic_defaults acceptance test

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
readme update

dimuon committed:
README update

dimuon committed:
README update

dimuon committed:
README update

dimuon committed:
README update

dimuon committed:
README update

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
fix deployment_basic_tags_test

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
fixes

dimuon committed:
fixes for ccs acc tests

dimuon committed:
add missing file

dimuon committed:
fix unit test

dimuon committed:
fixes for acc tests (WIP)

dimuon committed:
fix acceptance tests (WIP)

dimuon committed:
fix acceptance tests (WIP)

dimuon committed:
fix formatting, lint check, version (to 0.6.0)

dimuon committed:
fix examples

dimuon committed:
improve ES topology handling

dimuon committed:
Mark ES autoscaling, hot tier and other apps topology as required

dimuon committed:
fix unit test

dimuon committed:
make elasticsearch config required in "ec_deployment"

dimuon committed:
fix examples

dimuon committed:
fixes

dimuon committed:
fix

dimuon committed:
fix

dimuon committed:
fix to acc test; bring back parallel arg for acc test

dimuon committed:
fixes to acc tests

dimuon committed:
Make ES config required and its "plugins" Computed.

Reader returns empty ES config struct if backend responds with nil for config.

dimuon committed:
fix unit tests

dimuon committed:
fix unit tests

dimuon committed:
Update README

dimuon committed:
Temp fix - bypass 504 during file upload

renovate[bot] committed:
chore(deps): update peterjgrainger/action-create-branch action to v2.3.0 (#560)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Commit:
Update CHANGELOG.md and NOTICE

leejones committed:
docs: Add GCP PSC to Traffic Filter argument reference for type (#557)

The Traffic Filter resource docs include an example of the "GCP Private Service Connect type", but it was not listed in the argument reference for `type`.

Commit:
Update CHANGELOG.md and NOTICE

renovate[bot] committed:
fix(deps): update module github.com/hashicorp/terraform-plugin-sdk/v2 to v2.24.1 (#559)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Commit:
Update CHANGELOG.md and NOTICE

dimuon committed:
Update README with migration guide

dimuon committed:
make ES config optional

dimuon committed:
use plan modifies for ES and ES tier attributes

dimuon committed:
UseStateForUnknown for ES tier size

dimuon committed:
Remove some logic from legacyToNodeRoles - the tests pass without it

dimuon committed:
Remove ES tiers configs

dimuon committed:
nullify elasticsearch config if it's empty in config and in backend response

dimuon committed:
Remove APM topology and move its attributes to APM level

dimuon committed:
fix lint errors

dimuon committed:
Remove enterprise_search topology and move its attributes to enterprise_search level

dimuon committed:
Remove integrations_server topology and move its attributes to integrations_server level

tobio committed:
Utilise the migrate template API when applying changes to the deployment template id  (#547)

* Utilise the migrate template API when applying changes to the deployment template id

* Add changelog

* Fix acceptance tests

* Reset the instance_configuration_id so changes from the migration API are picked up

Commit:
Update CHANGELOG.md and NOTICE

dimuon committed:
fix acceptance tests and docs

dimuon committed:
Remove kibana topology and move its attributes to kibana level

dimuon committed:
Change plan modifier for size and few other attributes in ES tiers

dimuon committed:
WIP: fix syntax in migration acc tests

dimuon committed:
remove obsolete code

dimuon committed:
remove obsolete code

dimuon committed:
Plan modifier for node_types and remove check for node_types in migration to node_roles.

Remove check for node_types in state during migration to node_roles:
It looks like the API can handle the case now.

The case:
- state contains node_types
- the deployment version in state is >= 7.10 so node_roles can be used
- plan changes the deployment version

The previous implementation - use node_types,
so the version change and migration to node_roles should not be done in the same update call.

New behaviour - use node_roles.
So the deployment version change and migration to node_roles can be done in the same update call.
Proved by manual and acceptance testing (TestAccDeployment_pre_node_roles, TestAccDeployment_post_node_roles).

dimuon committed:
fix 2 acceptance tests

dimuon committed:
fix lint errors

dimuon committed:
plan modifier for node_roles + remove obsolete code

dimuon committed:
Merge remote-tracking branch 'upstream/master' into feature/530/ec-deployment

dimuon committed:
fix build - remove leftovers from non-migrated upstream changes

dimuon committed:
move apm tests to v2 package

dimuon committed:
move enterprise_search tests to v2

dimuon committed:
move integrations_server payload tests to v2

dimuon committed:
adhoc fix

dimuon committed:
fix panic during importing

dimuon committed:
improve comments + cosmetic improvement

dimuon committed:
fix acc test for import

dimuon committed:
move integrations_server read tests to v2

dimuon committed:
move kibana payload and read tests to v2

dimuon committed:
move observability payload and read tests to v2

dimuon committed:
move elasticsearch read tests to v2

dimuon committed:
move elasticsearch payload tests to v2

dimuon committed:
move deployment remote_cluster payload and traffic_filter tests to v2

dimuon committed:
move deployment parse creds tests to v2

dimuon committed:
(WIP) move deployment read test to v2

dimuon committed:
move deployment read test to v2

dimuon committed:
fix compiliation error

dimuon committed:
temp change to fix the branch build

dimuon committed:
(WIP) move deployment update payload test to v2

dimuon committed:
move deployment update payload test to v2

dimuon committed:
remove obsolete code

dimuon committed:
remove obsolete code

dimuon committed:
rearrange apm files

dimuon committed:
rearrange ec_deployment enterprise_search files

dimuon committed:
rearrange elasticsearch files

dimuon committed:
rearrange integrations_server files

dimuon committed:
rearrange ec_deployment kibana files

dimuon committed:
rearrange ec_deployment observability files

dimuon committed:
add missing file

dimuon committed:
refactor apm v2

dimuon committed:
refactor ec_deployment v2

dimuon committed:
refactor ec_deployment v2

dimuon committed:
Remove old ec_deployment implementation

dimuon committed:
fix an acc test error

dimuon committed:
ec_deployment_extension: fail update insted of warning

pascal-hofmann commented on README.md:
```diff
@@ -114,3 +120,110 @@ $ export EC_API_KEY="<apikey value>"
+- import existing resource using deployment ID, e.g `terraform import 'ec_deployment.test' <deployment_id>`
+- state upgrade that is performed by TF by calling the provider's API so no action is required from user perspective
+
+Currently the state upgrade functioanlity is still in development so importing existing resources is the recommended way to deal with existing TF states.
```
```suggestion
Currently the state upgrade functionality is still in development so importing existing resources is the recommended way to deal with existing TF states.
```

pascal-hofmann commented on ec/acc/deployment_autoscaling_test.go:
```diff
@@ -44,81 +44,63 @@ func TestAccDeployment_autoscaling(t *testing.T) {
 			// also disables ML
 			{
 				Config: cfgF(disableAutoscale),
 				// When disabling a tier the plan will be non empty on refresh
```
This comment is now a lie, right?

pascal-hofmann commented on ec/acc/testdata/deployment_autoscaling_2.tf:
```diff
@@ -9,43 +9,42 @@ resource "ec_deployment" "autoscaling" {
+      size        = "0g"
+      zone_count  = 1
+      autoscaling = {}
+    },
```
```suggestion
    }
```

pascal-hofmann commented on ec/acc/testdata/deployment_autoscaling_2.tf:
```diff
@@ -9,43 +9,42 @@ resource "ec_deployment" "autoscaling" {
+      size        = "0g"
+      zone_count  = 1
+      autoscaling = {}
+    },
```
```suggestion
    }
```

pascal-hofmann commented on ec/acc/testdata/deployment_autoscaling_2.tf:
```diff
@@ -9,43 +9,42 @@ resource "ec_deployment" "autoscaling" {
-    }
-    topology {
-      id         = "ml"
+    },
```
```suggestion
    }
```

pascal-hofmann commented on ec/acc/testdata/deployment_autoscaling_2.tf:
```diff
@@ -9,43 +9,42 @@ resource "ec_deployment" "autoscaling" {
-    }
-    topology {
-      id         = "warm"
+    },
```
```suggestion
    }
```

pascal-hofmann reviewed (approved):
LGTM apart from minor things. I did not do any real world tests with the new ec_deployment resource though.

pascal-hofmann commented on ec/acc/deployment_autoscaling_test.go:
```diff
@@ -44,81 +44,63 @@ func TestAccDeployment_autoscaling(t *testing.T) {
 			{
 				Config: cfgF(disableAutoscale),
 				// When disabling a tier the plan will be non empty on refresh
 				// since the topology block is present with size = "0g".
-				ExpectNonEmptyPlan: true,
+				// ExpectNonEmptyPlan: true,
```
This comment is now a lie, right?

dimuon replied:
I think yes, I'll remove the comment.

dimuon committed:
address PR review comments

dimuon committed:
improve test data

dimuon committed:
improve README

dimuon committed:
test deployment shutdown

dimuon committed:
small refactoring of create & update

dimuon committed:
remove unused code

dimuon committed:
improve acceptance test
